db/kvstore_versioned.go

```diff
@@ -47,4 +52,107 @@ type (
 		// SetVersion sets the version, and returns a KVStore to call Put()/Get()
 		SetVersion(uint64) KVStore
 	}
+
+	// KvWithVersion wraps the versioned DB implementation with a certain version
+	KvWithVersion struct {
+		db        VersionedDB
+		kvBase    KVStore
+		versioned map[string]bool // map of versioned buckets
+		version   uint64          // the current version
+	}
 )
+
+// Option sets an option
+type Option func(*KvWithVersion)
+
+func VersionedNamespaceOption(ns ...string) Option {
+	return func(k *KvWithVersion) {
+		k.versioned = make(map[string]bool)
+		for _, ns := range ns {
+			k.versioned[ns] = true
+		}
+	}
+}
+
+// NewKVStoreWithVersion implements a KVStore that can handle both versioned
+// and non-versioned namespace
+func NewKVStoreWithVersion(cfg Config, opts ...Option) *KvWithVersion {
+	db := NewBoltDBVersioned(cfg)
+	kv := KvWithVersion{
+		db:     db,
+		kvBase: db.Base(),
+	}
+	for _, opt := range opts {
+		opt(&kv)
+	}
+	return &kv
+}
+
+// Start starts the DB
+func (b *KvWithVersion) Start(ctx context.Context) error {
+	return b.kvBase.Start(ctx)
+}
+
+// Stop stops the DB
+func (b *KvWithVersion) Stop(ctx context.Context) error {
+	return b.kvBase.Stop(ctx)
+}
+
+// Put writes a <key, value> record
+func (b *KvWithVersion) Put(ns string, key, value []byte) error {
+	if b.versioned[ns] {
+		return b.db.Put(b.version, ns, key, value)
+	}
+	return b.kvBase.Put(ns, key, value)
```

Comment:
Using kvBase in kvWithVersion is strange, I would recommend splitting it up. for example:
```
KvVersioned interface{
    Version(string, []byte) (uint64, error)
    SetVersion(uint64) KVStore
    Plain() KVStore
}
```